RPC server tests

ErEmEy7 opened the pull request.

ErEmEy7 committed:
first server RpcTest

nepoklonov committed:
processRequest return type fixed; prepareArguments prettified

nepoklonov committed:
Methods.kt renamed to Method.kt

EvaOrange73 committed:
Method.kt and RpcController.kt moved;

ErEmEy7 committed:
Merge remote-tracking branch 'origin/dev' into tests

# Conflicts:
#	server/src/main/kotlin/rpc/Rpc.kt

ErEmEy7 committed:
shared gradle: implementation for Json
JsonVar.kt: json var for rpc/tests
Rpc.kt: Json -> json

ErEmEy7 committed:
ExampleServerController.kt: added test functions
ExampleTest.kt: added rpc tests

nepoklonov commented on shared/src/commonMain/kotlin/rpc/JsonVar.kt:
```diff
@@ -0,0 +1,8 @@
+
+import kotlinx.serialization.json.Json
+
+val json = Json {
```
Я бы переименовал файл в `Json.kt`

nepoklonov commented on server/src/test/kotlin/ExampleServerController.kt:
```diff
@@ -0,0 +1,112 @@
+class TestController : RpcController {
+
+    @Method(MethodType.GET)
+    fun getInt(arg1: Int, arg2: Int): Int {
```
Хочется, чтобы названия функций как-то отражали их суть

nepoklonov replied:
Возможно, чтобы избавиться от необходимости везде писать get/post в названиях функций, стоит сделать два контроллера: `TestGetController` и `TestPostController`

nepoklonov replied:
Ооо, а может, можно навесить на одну функцию сразу две аннотации? Я читал, что так можно делать, начиная с какой-то версии котлина. Правда, для этого нужно будет чуть-чуть доработать штуку, которая эти аннотации считывает

nepoklonov replied:
В любом случае, нужно отказаться от дублирования кода, самый простой вариант -- пусть одна из функций вызывает другую

ErEmEy7 replied:
Насчет 2-х контроллеров, функции же все равно по-разному называть, при вызове их в тесте, не совсем понял, что изменится

ErEmEy7 replied:
А вынести основную функцию, чтобы функции под аннотациями ее просто вызывали - это замечательно

ErEmEy7 replied:
Не получается вынести функцию за аннотацию, rpc выдает ошибку, что нет аннотации, хотя на основной функции аннотация есть

nepoklonov replied:
"Всё равно по-разному называть".
Да, есть такой момент, честно говоря, немного странный. У нас сейчас так всё устроено, что мы не можем в двух разных контроллерах назвать одинаково функции. Достаточно странное правило, которое, к слову, 1) хорошо бы проверять при запуске (чтобы сервер кидал ошибку, если мы пытаемся подключить два конфликтующих контроллера); 2) уметь как-то обходить. 

Во-первых, мб сделать аннотации на контроллер, типа:

```
@Route("users")
class UserController {
    @Method(MethodType.POST)
    fun getAll(): List<User> { ... }
}
```
Чтобы функция `getAll` соответствовала запросу `/api/users/getAll`.

Во-вторых, сделать возможность принудительного именования:

```
class UserController {
    @OverrideName("get-all-users")
    fun getAll(): List<User> { ... }
}
```
Чтобы функция `getAll` соответствовала запросу `/api/get-all-users`.

nepoklonov replied:
Вынести тело функций в отдельную, я имел в виду следующее:

```
class UserController {
    fun calculateSomething(arg1: Int) = ...
    
    @Method(MethodType.POST)
    fun postSomething(arg1: Int) = calculateSomething(arg1)

    @Method(MethodType.GET)
    fun getSomething(arg1: Int) = calculateSomething(arg1)
}
```

nepoklonov commented on server/src/test/kotlin/ExampleServerController.kt:
```diff
@@ -0,0 +1,112 @@
+
+    @Method(MethodType.GET)
+    fun getList(arg1: MutableList<Int>, arg2: MutableList<String>): MutableList<String> {
+        val ans: MutableList<String> = mutableListOf()
```
Лучше так, вместо шести строчек:
`return arg1.indices.flatMap {
    listOf(arg1[it].toString(), arg2[it])
}`

nepoklonov replied:
Я бы вообще какую-нибудь более понятную операцию делал, например, складывал бы эти листы, вот так: `return arg1.map { it.toString() } + arg2`

nepoklonov commented on server/src/test/kotlin/ExampleServerController.kt:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Method(MethodType.GET)
+    fun getMM(arg1: Map<Map<Int, String>, Map<String,Int>>, arg2: Map<Map<String, Int>, Map<Int,String>>): Map<Map<String, String>, Map<String, String>> {
```
Какая-то мутная функция. 
Я бы сделал где-нибудь в `commonMain` довольно общую функцию: 
```
fun <A, B, C> joinMaps(a: Map<A, B>, b: Map<B, C>): Map<A, C> {
    return a.mapNotNull { it.key to (b[it.value] ?: return@mapNotNull null) }.toMap()
}
```
А затем во всех таких функциях делал бы join этих мап. Как раз по типам подходит.

ErEmEy7 replied:
Ну да, а эта функция кристально понятная как будто............. 
Но хорошо, я ее вынесу просто в приватную функцию как и все остальные в контроллере (если она действительно нужна отдельным файлом, так еще и в shared, то я просто не понял зачем)

ErEmEy7 replied:
Ладно, я кстати вообще не понимаю, как работать с этой функцией

nepoklonov replied:
Функция `joinMaps`, которую я написал, имеет кристально ясную, на мой взгляд, суть. Она берёт две мапы и склеивает ключи первой и значения второй по совпадающим значениям первой и ключам второй. Иными словами, из мап: `{1: "a", 2: "b", 3: "c"}` и `{"a": "A", "b": "B", "c": "C"}` она делает мапу: `{1: "A", 2: "B", 3: "C"}`.

Её простой код был бы такой:

```
fun <A, B, C> joinMaps(a: Map<A, B>, b: Map<B, C>): Map<A, C> {
    return a.mapValues { b[it.value]!! }
}
```
Но он небезопасен: если мы дадим две мапы:
 `{1: "a", 2: "b", 3: "c"}` и `{"a": "A", "b": "B"}`, он кинет ошибку `NullPointerException`, потому что не найдёт ключа "c" во второй мапе. А чуть более сложный код, который я написал в изначальном комментарии, вернёт: `{1: "A", 2: "B"}`.

Как с ней работать -- вызывать на подходящих мапах, где значения первой совпадают с ключами второй.

nepoklonov commented on server/src/test/kotlin/ExampleServerController.kt:
```diff
@@ -0,0 +1,112 @@
+
+    @Method(MethodType.GET)
+    fun getPair(arg1: Pair<Int, Int>, arg2: Pair<Char, Char>): Pair<String, String> {
+        return Pair("${arg1.first}${arg2.first}", "${arg1.second}${arg2.second}")
```
Хочется какой-то более естественной логики

ErEmEy7 replied:
Изначально хотел val str: String = char1 + char2 , но так нельзя, поэтому решил не париться и ебнуть строчку 
А что там можно адекватного сделать

nepoklonov replied:
Я бы вернул пару из этих двух пар)

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,281 @@
+const val strArg1: String = "222"
+const val strArg2: String = "555"
+
+val listArg1: MutableList<Int> = mutableListOf(1,2,3,4,5)
```
`Ctrl + Alt + L`

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,281 @@
+val listArg2: MutableList<String> = mutableListOf("a","b","c","d","e")
+
+val mapArg1: Map<Int, String> = mapOf(1 to "one", 2 to "two")
+val mapArg2: Map<String, Int> = mapOf("three" to 3, "four" to 4)
```
В свете комментариев выше я бы вторую мапу сделал обратной к первой:
```
val mapArg1: Map<Int, String> = mapOf(1 to "one", 2 to "two", 3 to "three", 4 to "four")
val mapArg2: Map<String, Int> = mapOf("one" to 1, "two" to 2, "three" to 3, "four" to 4)
```

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,281 @@
+    @Test
+    fun `send two lists and receive their combination by GET`() {
+        withTestApplication(Application::test) {
+            handleRequest(HttpMethod.Get, "/api/getList?arg1=$listArg1&arg2=${json.encodeToString(listArg2)}").apply {
```
А как так получилось, что один аргумент закодирован, а другой -- нет?
Думаю, это случайность, если оно работает. Чисто совпадение, что `toString` в данном случае работает так же как и `encodeToString`.

nepoklonov replied:
Чтобы избежать таких ошибок, я бы:
1. Сделал обёртку для всех этих тестов, чтобы можно было писать, типа:
```
@Test
fun `send two lists`() = testRpc(
    url = "/api/getList", 
    type = MethodType.GET
    body = build {
        let("arg1", listArg1)
        let("arg2", listArg2)
    }
    expected = TestController().getList(listArg1, listArg2)
)
```
Сразу же становится понятно, что проверяется. Вся лишняя логика выносится в общую функцию

nepoklonov replied:
2. Добавил функциональность в `build` (у тебя она называется `builder`, лучше использовать глагол), чтобы можно было билдить не только пост-запросы, но и гет-запросы.

ErEmEy7 replied:
Это не ошибка, там один список интовый, который не нужно енкодить, а второй - стринговый 
Насчет обертки, не совсем понял, это должна быть отдельная функция для каждого теста?

nepoklonov replied:
Если енкодить выборочно, то может получиться такая проблема: ты меняешь тип данных с `Int` на `String`, допустим, а encode забываешь добавить. Классическая проблема, возникающая от отсутствия унификации.

Обёртку хочется написать такую, чтобы все тесты её вызывали, как в примере выше. Сейчас каждый твой тест вызывает `withTestApplication` и ещё кучу строчек. Хочется сделать более простое тело функции для тестов.

nepoklonov reviewed (commented):
Небольшие замечания, в основном, по естественности логики функций в контроллере и по дублированию кода в самом тесте.

ErEmEy7 committed:
ExampleServerController.kt: splited on GetServerController.kt and PostServerController.kt and deleted. The functions are placed in the Implementation.kt
ExampleTest.kt: Added a function for building get requests. Added a wrapper for creating tests. All tests have been rewritten.
Json.kt: renamed from JsonVar.kt
ServerInterface.kt: An interface has been created containing a description of functions from Implementation
Utils.kt: fun MapJoinFun removed from ExampleServerController.kt to Utils.kt
TestApplication.kt: Split TestController class on PostServerController class and GetServerController class

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,246 @@
+val mmArg1: Map<Map<Int, String>, Map<String, Int>> = mapOf(mapArg1 to mapArg2)
+val mmArg2: Map<Map<String, Int>, Map<Int, String>> = mapOf(mapArg2 to mapArg1)
+
+val pairArg1: Pair<Int, Int> = Pair(1, 2)
```
По-котлиновски лучше писать так:
`val pairArg1: Pair<Int, Int> = 1 to 2`

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,246 @@
+    val map: MutableMap<String, String> = mutableMapOf()
+    var getUrl: String = ""
+    
+    inline fun <reified T> setPost(name: String, value: T) {
```
Предлагаю разделить ответственность между `Body` и `testRpc` таким образом. Пусть `Body` собирает мапу аргументов `Map<String, String>`, а `testRpc` обрабатывает эту мапу, как ей нужно. Либо вызывает `setBody`, либо дописывает аргументы к `url` через `?` и `&`.

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,246 @@
+    }
+}
+
+inline fun <reified T> testRpc(url: String, type: HttpMethod, body: Map<String, String>? = null, expected: T) =
```
Лучше сделать `body: Map<String, String> = mapOf()` вместо `body: Map<String, String>? = null`. А ещё лучше, с учётом комментария выше, `body: Body`

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,246 @@
+import io.ktor.application.*
+import io.ktor.http.*
+import io.ktor.http.HttpMethod.Companion.Get
```
Возможно, лучше использовать наши `MethodType` вместо кторовских `HttpMethod`

ErEmEy7 replied:
Для handleRequest-а нужен HttpMethod

nepoklonov replied:
Согласен. Но для нашей обёртки -- нет. Предлагаю в обёртку передавать `MethodType`, а она пусть делает из него `HttpMethod` при помощи `HttpMethod.valueOf(...)`.

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,246 @@
+    )
+
+    @Test
+    fun `send one sealed class instance and receive him num by POST`() = testRpc(
```
`him` -> `its`

nepoklonov commented on server/src/test/kotlin/Utils.kt:
```diff
@@ -0,0 +1,2 @@
+fun <A, B, C> mapJoinFun(arg1: Map<A, B>, arg2: Map<B, C>): Map<A, C> =
```
Лучше сделать её сигнатуру `Map<A, B>.joinWith(other: Map<B, C>): Map<A, C>`.

ErEmEy7 replied:
Наверное, про это нужно что-то почитать, потому что не особо понятно

nepoklonov commented on server/src/test/kotlin/ServerInterface.kt:
```diff
@@ -0,0 +1,14 @@
+interface ServerInterface {
```
Напрягает очень общее название для этого интерфейса (кстати, лучше не использовать `interface` в названиях интерфейсов без необходимости), а также всех, кто от него наследуется. Хочется, чтобы название отражало суть (тестовый контроллер для проверки RPC).

nepoklonov commented on server/src/test/kotlin/Implementation.kt:
```diff
@@ -0,0 +1,27 @@
+class Implementation : ServerInterface {
```
Тут тоже (см. комментарий к `ServerInterface`) проблема с очень общим названием

nepoklonov commented on server/src/test/kotlin/Implementation.kt:
```diff
@@ -0,0 +1,27 @@
+
+    override fun mapJoin(arg1: Map<Int, String>, arg2: Map<String, Int>): Map<Int, Int> = mapJoinFun(arg1, arg2)
+
+    override fun mmJoin(arg1: Map<Map<Int, String>, Map<String, Int>>, arg2: Map<Map<String, Int>, Map<Int, String>>):
```
Зачем так сложно, если можно просто вызвать `mapJoinFun(arg1, arg2)`?

ErEmEy7 replied:
Чтобы все функции, которые используются в контроллере, описывались в этом самом контроллере
Логично же

nepoklonov replied:
Не понял ответ. Перефразирую вопрос.
У нас в контроллере есть две функции: `mapJoin` и `mmJoin`. Первая -- простая и понятная, вторая -- странная и сложная. Вопрос следующий: почему бы не сделать у `mmJoin` такую же логику, как и у `mapJoin`?? Там как раз мапы подходящие по типам.

nepoklonov commented on server/src/test/kotlin/Implementation.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    override fun pairOfPair(arg1: Pair<Int, Int>, arg2: Pair<Char, Char>): Pair<Pair<Int, Int>, Pair<Char, Char>> =
+        Pair(arg1, arg2)
```
Исправить на `arg1 to arg1`.

nepoklonov commented on server/src/test/kotlin/ExampleTest.kt:
```diff
@@ -0,0 +1,246 @@
+val pairArg1: Pair<Int, Int> = Pair(1, 2)
+val pairArg2: Pair<Char, Char> = Pair('a', 'b')
+
+enum class EnumArguments(val par: Int) {
```
Лучше эти тестовые классы вынести в `commonMain`.

nepoklonov reviewed (commented):
Практически идеально. Хорошо бы немного подправить по мелочам, в том числе, архитектурным, чтобы чистота кода была на высшем уровне.

ErEmEy7 committed:
EnumArgumentsForRPCTests.kt: enum class from ExampleTest.kt
SealedArgumentsForRPCTests.kt: sealed class from ExampleTest.kt
ExampleTest.kt: enum and sealed classes removed; buildGet and Body changed;
ImplementationFunsForRPCTests.kt: renamed from Implementation.kt
TestRpcController.kt: renamed from ServerInterface.kt
Utils.kt: mapJoinFun reworked

nepoklonov committed:
testRPC and builders prettified

nepoklonov committed:
Naming improved, locations fixed
